src/mlnet/Telemetry/Telemetry.cs

```diff
@@ -1,144 +1,239 @@
             {
-                Debug.Fail(e.ToString());
+                eventProperties["Exception"] = GetSanitizedExceptionStr(ex);
             }
-        }
 
-        private Dictionary<string, double> GetEventMeasures(IDictionary<string, double> measurements)
-        {
-            Dictionary<string, double> eventMeasurements = new Dictionary<string, double>();
-            if (measurements != null)
+            if (properties != null)
             {
-                foreach (KeyValuePair<string, double> measurement in measurements)
+                foreach (KeyValuePair<string, string> property in properties)
                 {
-                    eventMeasurements[measurement.Key] = measurement.Value;
+                    if (property.Value != null)
+                    {
+                        eventProperties[property.Key] = property.Value;
+                    }
                 }
             }
-            return eventMeasurements;
+
+            return eventProperties;
         }
 
-        private Dictionary<string, string> GetEventProperties(IDictionary<string, string> properties)
+        private static bool CheckFirstTimeUse()
         {
-            if (properties != null)
+            using (IFirstTimeUseNoticeSentinel firstTimeUseNoticeSentinel = new FirstTimeUseNoticeSentinel())
             {
-                var eventProperties = new Dictionary<string, string>(_commonProperties);
-                foreach (KeyValuePair<string, string> property in properties)
+                var firstTimeUse = !firstTimeUseNoticeSentinel.Exists();
+
+                if (firstTimeUse)
                 {
-                    eventProperties[property.Key] = property.Value;
+                    Console.WriteLine(
+@"Welcome to the ML.NET CLI!
+--------------------------
+Learn more about ML.NET CLI: https://aka.ms/mlnet-cli
+Use 'mlnet --help' to see available commands or visit: https://aka.ms/mlnet-cli-docs
+
+Telemetry
+---------
+The ML.NET CLI tool collects usage data in order to help us improve your experience.
+The data is anonymous and doesn't include personal information or data from your datasets.
+You can opt-out of telemetry by setting the MLDOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.
+
+Read more about ML.NET CLI Tool telemetry: https://aka.ms/mlnet-cli-telemetry
+");
+
+                    firstTimeUseNoticeSentinel.CreateIfNotExists();
+
+                    return true;
                 }
-                return eventProperties;
-            }
-            else
-            {
-                return _commonProperties;
+
+                return false;
             }
         }
+
+        /// <summary>
+        /// Get exception string to log. Exclude the exception message, as it
+        /// may contain PII.
+        /// </summary>
+        private static string GetSanitizedExceptionStr(Exception ex)
+        {
+            return $@"{ex.GetType()}
+{ex.StackTrace}";
```

Comment:
Thinking out loud..  if the exception somehow happened in user code, we will get their names as part of the stack trace.  Maybe also sanitize the stack trace itself to only include stuff from Microsoft.* namespace?

Reply:
Interesting. Right now, I don't think this is possible, of course. Are you thinking of this as a defensive move against future development / capabilities in this repository?